File: enter.pollinations.ai/scripts/user-pipeline/shared/d1.py
import json
import subprocess
import sys
from pathlib import Path


def ensure_safe_env(env: str) -> str:
    if env != "staging":
        print(
            f"❌ Unsupported env: {env}. This branch is locked to staging and cannot write to production.",
            file=sys.stderr,
        )
        sys.exit(1)
    return env


def run_d1_query(query: str, env: str = "staging") -> list[dict] | None:
    """Run a D1 query and return results."""
    env = ensure_safe_env(env)
    cmd = [
        "npx",
        "wrangler",
        "d1",
        "execute",
        "DB",
        "--remote",
        "--env",
        env,
        "--command",
        query,
        "--json",
    ]

    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            cwd=Path(__file__).resolve().parents[3],
            timeout=120,
        )

        if result.returncode != 0:
            print(f"❌ D1 query failed: {result.stderr}", file=sys.stderr)
            return None

        data = json.loads(result.stdout)
        return data[0].get("results", [])

    except subprocess.TimeoutExpired:
        print("❌ D1 query timed out", file=sys.stderr)
        return None
    except (json.JSONDecodeError, KeyError, IndexError) as error:
        print(f"❌ Failed to parse D1 response: {error}", file=sys.stderr)
        return None

File: enter.pollinations.ai/scripts/user-pipeline/shared/d1.ts
/**
 * Thin wrapper around `wrangler d1 execute` for running SQL against a remote D1 database.
 * queryD1 returns rows; executeD1 runs mutations and returns success/failure.
 */

import { execFileSync } from "node:child_process";

type Environment = "staging" | "production";
type D1Row = Record<string, string | number | null>;

interface D1JsonResponse {
    results?: D1Row[];
}

function buildWranglerArgs(env: Environment, sql: string): string[] {
    return [
        "wrangler",
        "d1",
        "execute",
        "DB",
        "--remote",
        "--env",
        env,
        "--command",
        sql,
        "--json",
    ];
}

export function queryD1(env: Environment, sql: string): D1Row[] {
    const output = execFileSync("npx", buildWranglerArgs(env, sql), {
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
        maxBuffer: 100 * 1024 * 1024,
        cwd: process.cwd(),
    });

    const data = JSON.parse(output) as D1JsonResponse[];
    return data[0]?.results || [];
}

export function executeD1(env: Environment, sql: string): boolean {
    try {
        execFileSync("npx", buildWranglerArgs(env, sql), {
            encoding: "utf-8",
            stdio: ["pipe", "pipe", "pipe"],
            maxBuffer: 20 * 1024 * 1024,
            cwd: process.cwd(),
        });
        return true;
    } catch (error) {
        console.error(
            "❌ D1 mutation failed:",
            error instanceof Error ? error.message : String(error),
        );
        return false;
    }
}
